galaxy_ng/app/tasks/index_registry.py

```diff
@@ -0,0 +1,160 @@
+                    return
+
+            except exceptions.ObjectDoesNotExist:
+                raise CouldNotCreateContainerError(
+                    container_data['name'],
+                    error=_(
+                        "A remote container with this name already exists, "
+                        "but is not associated with any registry.")
+                )
+
+        else:
+            raise CouldNotCreateContainerError(
+                container_data['name'],
+                error=_("A local container with this name already exists.")
+            )
+
+    except exceptions.ObjectDoesNotExist:
+        # If no distributions match the selected container, create one.
+        request = _get_request(request_data)
+        serializer = serializers.ContainerRemoteSerializer(
+            data={
+                "name": container_data['name'],
+                "upstream_name": container_data['name'],
+                "registry": str(registry_pk)
+            }, context={"request": request}
+        )
+
+        try:
+            serializer.is_valid(raise_exception=True)
+        except serializers.ValidationError as e:
+            CouldNotCreateContainerError(
+                container_data['name'],
+                error=str(e)
+            )
+        serializer.create(serializer.validated_data)
+
+        distro = container_models.ContainerDistribution.objects.get(
+            base_path=container_data['name'])
+
+        _update_distro_readme_and_description(distro, container_data)
+
+
+def index_execution_environments_from_redhat_registry(registry_pk, request_data):
+    registry = models.ContainerRegistryRemote.objects.get(pk=registry_pk)
+    remotes = []
+
+    query = {
+        "filter": "build_categories=in=('Automation Execution Environment')",
+        "page": 0,
+        "sort_by": "creation_date[asc]"
+    }
+
+    # Iterate through the pages on the API response and add the relevant information
+    # about each container repository.
+    while True:
+        url = CATALOG_API + "?" + urlencode(query, quote_via=quote)
+        downloader = registry.get_downloader(url=url)
+        download_result = downloader.fetch()
+        with open(download_result.path) as fd:
+            data = json.load(fd)
+            remotes = remotes + _parse_catalog_repositories(data)
+            if len(data['data']) == data['page_size']:
+                query['page'] += 1
+            else:
+                break
+
+    for remote in remotes:
+        # create a subtask for each remote, so that if one fails, we can throw a usable error
+        # message for the user to look at and prevent the rest of the repositories from failing.
+        dispatch(
+            create_or_update_remote_container,
+            kwargs={
+                "container_data": remote,
+                "registry_pk": registry.pk,
+                "request_data": request_data
+            },
+            exclusive_resources=['/pulp/api/v3/distributions/']
```

Comment:
for `exclusive_resources` (or `resources` in the past), iirc we send a list of model objects or list of string pk of a model object https://docs.pulpproject.org/pulpcore/plugins/api-reference/tasking.html#pulpcore.plugin.tasking.dispatch

Is an endpoint considered a resource that it looks up to find objects? and would it lock every distribution in the database? If you need to lock all distros assocated to containers, maybe you could send it something like `models.ContainerDistribution.objects.all()`

Reply:
I think I need to pass a distribution instance here. Fabricio had a note about this in IRC.

Reply:
when you create distributions you need to put a lot on all distributions so that they can validate that the base path doesn't overlap

Reply:
> I think I need to pass a distribution instance here

I assumed this was an instance of type models.*Distribution rather than a url, I did a quick search in pulpcore and pulp_ansible for resources|exclusive_resources|shared_resources and they seem to all pass database objects. But the url may just be a pattern I don't know.

I _thought_ the only reason it allowed a string to be passed in, was in case it was a pk

Reply:
this is correct as-is per pulp folks!